Adding Vulnerability Assessment cmdlets on Server and Managed Instance

talhers opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description

These cmdlets are meant for setting Vulnerability Assessment settings on server and managed instance, and they are the same as the Vulnerability Assessment settings cmdlets on database and managed database.

Design review for Vulnerability Assessment cmdlets can be found at:
[https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/236](https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/236)



## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (Microsoft internal partners only)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/help-generation.md)

talhers commented:
@yoavfr @yaakoviyun @jaredmoo

talhers commented on src/Sql/Sql/VulnerabilityAssessment/Cmdlet/SqlVulnerabilityAssessmentAtpCmdletBase.cs:
```diff
@@ -55,7 +55,7 @@ protected override M ApplyUserInputToModel(M model)
             // Verify that ATP is ON - if ATP is off the GetServerAdvancedThreatProtectionPolicy() will throw a not found exception
             var sqlAdvancedThreatProtectionAdapter = new SqlAdvancedThreatProtectionAdapter(DefaultProfile.DefaultContext);
-            if (GetResourceTypeVaAppliesTo() == ApplyToType.Database)
+            if (GetResourceTypeVaAppliesTo() == ApplyToType.Database || GetResourceTypeVaAppliesTo() == ApplyToType.Server)
```
Yes, because the policy object returned is different, and the error message is different

talhers commented on src/Sql/Sql/VulnerabilityAssessment/Cmdlet/VulnerabilityAssessmentSettings/Server/GetAzureSqlServerVulnerabilityAssessmentSettings.cs:
```diff
@@ -0,0 +1,88 @@
+        [Parameter(ParameterSetName = GetByParentResourceParameterSet,
+            Mandatory = false,
+            ValueFromPipeline = true,
+            HelpMessage = "The database object to get Vulnerability Assessment settings for")]
```
Fixed

talhers commented on src/Sql/Sql.Test/ScenarioTests/VulnerabilityAssessmentMiTests.ps1:
```diff
@@ -23,6 +23,211 @@ function Test-VulnerabilityAssessmentManagedInstanceWithAtpOffTest
 
+	try
+	{
+		$atpOffErrorMessage = "Managed Instance Advanced Threat Protection is not defined. Please run Enable-AzSqlInstanceAdvancedDataSecurity first."
```
Fixed

talhers commented on src/Sql/Sql/VulnerabilityAssessment/Services/BaseSqlVulnerabilityAssessmentAdapter.cs:
```diff
@@ -56,22 +56,34 @@ public BaseSqlVulnerabilityAssessmentAdapter(IAzureContext context)
+                    var vaManagedInstanceSettings = VulnerabilityAssessmentCommunicator.GetManagedInstanceVulnerabilityAssessmentSettings(resourceGroup, serverName);
+                    return ConvertSettingsToModel(resourceGroup, vaManagedInstanceSettings.RecurringScans, vaManagedInstanceSettings.StorageContainerPath);
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(applyToType), applyToType, null);
```
Fixed it to something more user-friendly

talhers commented on src/Sql/Sql/VulnerabilityAssessment/Cmdlet/SqlVulnerabilityAssessmentAtpCmdletBase.cs:
```diff
@@ -80,10 +80,13 @@ protected override M ApplyUserInputToModel(M model)
             {
-                WriteWarning(Resources.VulnerabilityAssessmentOperationIsNotAvailableOnMasterDatabase);
-                return;
+                if (GetDatabaseName().ToLower() == "master")
```
In managed instance we're not getting master at all. I changed the condition to check only in regular databases

MiYanni commented:
@vladimir-shcherbakov Please reassign to someone else. Thanks.

vladimir-shcherbakov commented:
@talhers Here Azure/azure-powershell-cmdlet-review-pr#78 I only found design review of the``` *SqlDatabaseVulnerabilityAssessment*```  cmdlets.
In the PR I found implementation of the cmdlts:
```
*SqlInstanceVulnerabilityAssessment*
*SqlServerVulnerabilityAssessment*
```
And .md files of the cmdlets
```
*SqlDatabaseVulnerabilityAssessment*
*SqlInstanceDatabaseVulnerabilityAssessment*
*SqlInstanceVulnerabilityAssessment*
```
How would you explain such a variety?

yoavfr commented:
> @talhers Here [Azure/azure-powershell-cmdlet-review-pr#78](https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/78) I only found design review of the` *SqlDatabaseVulnerabilityAssessment*` cmdlets.
> In the PR I found implementation of the cmdlts:
> 
> ```
> *SqlInstanceVulnerabilityAssessment*
> *SqlServerVulnerabilityAssessment*
> ```
> And .md files of the cmdlets
> 
> ```
> *SqlDatabaseVulnerabilityAssessment*
> *SqlInstanceDatabaseVulnerabilityAssessment*
> *SqlInstanceVulnerabilityAssessment*
> ```
> How would you explain such a variety?

@vladimir-shcherbakov 
The overall design for database vulnerability assessment applies with a minor change (no database parameter is provided) to the server level cmdlets. Everything else is identical.
The same is true for Managed Instance where everything is identical except for the name of the cmdlet and the reference inside to a differently named SDK (ManagedInstance instead of Server).
The help files have been altered for SqlDatabaseVulnerabilityAssessment since we now support server level policies (as opposed to database level) and this is the path that we recommend customers use.

vladimir-shcherbakov commented:
@talhers 
You need to go through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) for every cmdlet with public API changes.

vladimir-shcherbakov commented:
@talhers  Please fix Static Analysis issues: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell/10680/artifact/artifacts/BreakingChangeIssues.csv/*view*/

vladimir-shcherbakov commented:
@talhers Since the [design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/236) has introduced only the  `AzSqlInstanceVulnerabilityAssessmentSettings` and `AzSqlServerVulnerabilityAssessmentSettings`  subset of cmdlets - to make it clearer please make sure no other cmdlets are modified in the PR.

maddieclayton commented:
@talhers Ping on the comments above.

talhers commented on src/Sql/Sql/VulnerabilityAssessment/Cmdlet/VulnerabilityAssessmentSettings/ManagedInstance/ClearSqlManagedInstanceVulnerabilityAssessmentSettings.cs:
```diff
@@ -0,0 +1,89 @@
+    [Cmdlet("Clear", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "SqlInstanceVulnerabilityAssessmentSettings", SupportsShouldProcess = true, DefaultParameterSetName = GetByNameParameterSet), OutputType(typeof(ManagedInstanceVulnerabilityAssessmentSettingsModel))]
+    public class ClearSqlManagedInstanceVulnerabilityAssessmentSettings : ClearSqlVulnerabilityAssessmentSettingsBase
+    {
+        /// <summary>
```
managed instance cmdlet classes are named in the same manner, for example: GetAzureSqlManagedInstance

talhers commented on src/Sql/Sql/VulnerabilityAssessment/Cmdlet/VulnerabilityAssessmentSettings/ManagedInstance/GetAzureSqlManagedInstanceVulnerabilityAssessmentSettings.cs:
```diff
@@ -0,0 +1,89 @@
+    /// </summary>
+    [Cmdlet("Get", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "SqlInstanceVulnerabilityAssessmentSettings", DefaultParameterSetName = GetByNameParameterSet), OutputType(typeof(ManagedInstanceVulnerabilityAssessmentSettingsModel))]
+    public class GetAzureSqlManagedInstanceVulnerabilityAssessmentSettings : GetAzureSqlVulnerabilityAssessmentSettingsBase
+    {
```
managed instance cmdlet classes are named in the same manner, for example: GetAzureSqlManagedInstance

talhers commented on src/Sql/Sql/VulnerabilityAssessment/Cmdlet/VulnerabilityAssessmentSettings/ManagedInstance/UpdateAzureSqlManagedInstanceVulnerabilityAssessmentSettings.cs:
```diff
@@ -0,0 +1,62 @@
+    /// </summary>
+    [Cmdlet("Update", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "SqlInstanceVulnerabilityAssessmentSettings", SupportsShouldProcess = true, DefaultParameterSetName = UpdateSettingsWithStorageAccountNameParameterSet), OutputType(typeof(ManagedInstanceVulnerabilityAssessmentSettingsModel))]
+    public class UpdateAzureSqlManagedInstanceVulnerabilityAssessmentSettings : UpdateAzureSqlVulnerabilityAssessmentSettingsBase
+    {
```
Look at my previous answers, I follow the same naming convention as managed instance cmdlets

talhers commented on src/Sql/Sql/VulnerabilityAssessment/Cmdlet/VulnerabilityAssessmentSettings/Server/GetAzureSqlServerVulnerabilityAssessmentSettings.cs:
```diff
@@ -0,0 +1,88 @@
+    /// </summary>
+    [Cmdlet("Get", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "SqlServerVulnerabilityAssessmentSettings"), OutputType(typeof(ServerVulnerabilityAssessmentSettingsModel))]
+    public class GetAzureSqlServerVulnerabilityAssessmentSettings : GetAzureSqlVulnerabilityAssessmentSettingsBase
+    {
```
Look at my previous answers, I follow the same naming convention as managed instance cmdlets

talhers commented on src/Sql/Sql/VulnerabilityAssessment/Cmdlet/VulnerabilityAssessmentSettings/Server/UpdateAzureSqlServerVulnerabilityAssessmentSettings.cs:
```diff
@@ -0,0 +1,62 @@
+    /// </summary>
+    [Cmdlet("Update", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "SqlServerVulnerabilityAssessmentSettings", SupportsShouldProcess = true, DefaultParameterSetName = UpdateSettingsWithStorageAccountNameParameterSet), OutputType(typeof(DatabaseVulnerabilityAssessmentSettingsModel))]
+    public class UpdateAzureSqlServerVulnerabilityAssessmentSettings : UpdateAzureSqlVulnerabilityAssessmentSettingsBase
+    {
```
Look at my previous answers, I follow the same naming convention as managed instance cmdlets

talhers commented:
> @talhers Since the [design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/236) has introduced only the `AzSqlInstanceVulnerabilityAssessmentSettings` and `AzSqlServerVulnerabilityAssessmentSettings` subset of cmdlets - to make it clearer please make sure no other cmdlets are modified in the PR.

I removed all modifications to other cmdlets

vladimir-shcherbakov commented:
@talhers Ping on the comments above.

talhers commented on src/Sql/Sql/VulnerabilityAssessment/Cmdlet/VulnerabilityAssessmentSettings/ManagedInstance/ClearSqlManagedInstanceVulnerabilityAssessmentSettings.cs:
```diff
@@ -0,0 +1,102 @@
+    /// Clears the Vulnerability Assessment on a specific managed instance.
+    /// </summary>
+    [Cmdlet("Clear", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "SqlInstanceVulnerabilityAssessmentSettings", DefaultParameterSetName = ClearByNameParameterSet, SupportsShouldProcess = true), OutputType(typeof(ManagedInstanceVulnerabilityAssessmentSettingsModel))]
+    public class ClearSqlManagedInstanceVulnerabilityAssessmentSettings : ClearSqlVulnerabilityAssessmentSettingsBase
```
It is handled in the base class - AzureRMCmdlet

talhers commented on src/Sql/Sql/VulnerabilityAssessment/Cmdlet/VulnerabilityAssessmentSettings/ManagedInstance/UpdateAzureSqlManagedInstanceVulnerabilityAssessmentSettings.cs:
```diff
@@ -0,0 +1,83 @@
+    /// </summary>
+    [Cmdlet("Update", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "SqlInstanceVulnerabilityAssessmentSettings", SupportsShouldProcess = true, DefaultParameterSetName = UpdateSettingsWithStorageAccountNameParameterSet), OutputType(typeof(ManagedInstanceVulnerabilityAssessmentSettingsModel))]
+    public class UpdateAzureSqlManagedInstanceVulnerabilityAssessmentSettings : UpdateAzureSqlVulnerabilityAssessmentSettingsBase
+    {
```
It is handled in the base class - AzureRMCmdlet

talhers commented on src/Sql/Sql/VulnerabilityAssessment/Cmdlet/VulnerabilityAssessmentSettings/Server/ClearSqlServerVulnerabilityAssessmentSettings.cs:
```diff
@@ -0,0 +1,102 @@
+    /// </summary>
+    [Cmdlet("Clear", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "SqlServerVulnerabilityAssessmentSettings", DefaultParameterSetName = ClearByNameParameterSet, SupportsShouldProcess = true), OutputType(typeof(ServerVulnerabilityAssessmentSettingsModel))]
+    public class ClearSqlServerVulnerabilityAssessmentSettings : ClearSqlVulnerabilityAssessmentSettingsBase
+    {
```
It is handled in the base class - AzureRMCmdlet

vladimir-shcherbakov replied:
Could you please link me to the handler? I'm looking for something like this: 
```Powershell
[Cmdlet(VerbsCommon.Remove, "Foo", SupportsShouldProcess = true)]
public class RemoveFoo
{
    ...
    public override void ExecuteCmdlet()
    {
        if (ShouldProcess(target, actionMessage))
        {
            // make changes
        }
    }
}
```

talhers replied:
The handler is implemented in assembly Microsoft.Azure.PowerShell.Common
in class AzurePSCmdlet:

```powershell
protected virtual void ConfirmAction(string processMessage, string target, Action action)
{
	if (_qosEvent != null)
	{
		_qosEvent.PauseQoSTimer();
	}
	if (base.ShouldProcess(target, processMessage))
	{
		if (_qosEvent != null)
		{
			_qosEvent.ResumeQosTimer();
		}
		action();
	}
}
```


(class AzureRMCmdlet calls method ConfirmAction)

vladimir-shcherbakov replied:
@talhers  The `ConfirmAction` is a [helper method](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/should-process-confirm-impact.md#additional-common-methods) that you may use to properly implement confirmation in the implementation of ProcessRecord()..

Only your cmdlet knows  which action should be executed in case of confirmation or `-Whatif`.

Please take a look at how the helper method is used in a random cmdlet from our repo: https://github.com/Azure/azure-powershell/blob/master/src/Network/Network/ApplicationGateway/RemoveAzureApplicationGatewayCommand.cs#L56

talhers replied:
In Sql module, all cmdlet classes inherit from AzureSqlCmdletBase, and implement PersistChanges method.
PersistChanges is called from [AzureSqlCmdletBase.ExecuteCmdlet](https://github.com/Azure/azure-powershell/blob/master/src/Sql/Sql/Common/AzureSqlCmdletBase.cs#L119
) if ConfirmAction returns true.


For example, cmdlet [NewAzureSqlDatabase ](https://github.com/Azure/azure-powershell/blob/master/src/Sql/Sql/Database/Cmdlet/NewAzureSqlDatabase.cs)- uses "SupportsShouldProcess = true", but doesn't directly use ConfirmAction method, because it is used in ExecuteCmdlet in base class AzureSqlCmdletBase.

vladimir-shcherbakov replied:
Now I see. Thanks.

talhers commented on src/Sql/Sql/VulnerabilityAssessment/Cmdlet/VulnerabilityAssessmentSettings/Server/UpdateAzureSqlServerVulnerabilityAssessmentSettings.cs:
```diff
@@ -0,0 +1,83 @@
+    /// </summary>
+    [Cmdlet("Update", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "SqlServerVulnerabilityAssessmentSettings", SupportsShouldProcess = true, DefaultParameterSetName = UpdateSettingsWithStorageAccountNameParameterSet), OutputType(typeof(DatabaseVulnerabilityAssessmentSettingsModel))]
+    public class UpdateAzureSqlServerVulnerabilityAssessmentSettings : UpdateAzureSqlVulnerabilityAssessmentSettingsBase
+    {
```
It is handled in the base class - AzureRMCmdlet

vladimir-shcherbakov commented:
@talhers Please resolve the merge conflicts and address the comment above.

vladimir-shcherbakov commented on src/Sql/Sql/help/Az.Sql.md:
```diff
@@ -167,9 +173,6 @@ Gets Advanced Data Security policy of a managed instance.
 
-### [Get-AzSqlInstanceDatabaseBackupShortTermRetentionPolicy](Get-AzSqlInstanceDatabaseBackupShortTermRetentionPolicy.md)
-Gets a backup short term retention policy.
-
```
Why was this removed?

talhers replied:
I had a merge conflict where I added this line and it did not exist in the main repo. I guess it was removed by someone else, so I removed it too. This is not one of my cmdlets, so I will not add something owned by someone else

vladimir-shcherbakov commented on src/Sql/Sql/help/Clear-AzSqlInstanceVulnerabilityAssessmentSettings.md:
````diff
@@ -0,0 +1,156 @@
+            -Name "ManagedInstance01" `
+            | Clear-AzSqlInstanceVulnerabilityAssessmentSettings
+```
+
````
Please add samples output

talhers replied:
Clear-AzSqlInstanceVulnerabilityAssessmentSettings doesn't return anything on success

vladimir-shcherbakov replied:
if it doesn't - you may want to consider using `bool` for the output type of the [`Clear-AzSqlInstanceVulnerabilityAssessmentSettings`](https://github.com/Azure/azure-powershell/pull/8413/files?file-filters%5B%5D=.cs#diff-632902af0776c8292c91cb56a243ae9dR26) and [`Clear-AzSqlServerVulnerabilityAssessmentSettings`](https://github.com/Azure/azure-powershell/pull/8413/files?file-filters%5B%5D=.cs#diff-10494e091851c462e9f1cd916383f4abR26) cmdlets and add `-PassThru` parameter.

talhers replied:
These cmdlets are consistent with our other cmdlets (for databases and managed databases) so we will keep it as it is for now. We might change that in the future for all our cmdlets.

vladimir-shcherbakov replied:
It's not clear what you mean by the **consistency**  
* if the other cmdlets don't use their output types - you need to deprecate them following the document: https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-attribute-help.md#cmdletoutputbreakingchangeattribute. 
* else - please use the output types to return anything on success and update corresponding help files samples.

vladimir-shcherbakov commented on src/Sql/Sql/help/Clear-AzSqlServerVulnerabilityAssessmentSettings.md:
````diff
@@ -0,0 +1,156 @@
+            -ServerName "Server01" `
+            | Clear-AzSqlServerVulnerabilityAssessmentSettings
+```
+
````
Please add samples output.

talhers replied:
Clear-AzSqlServerVulnerabilityAssessmentSettings doesn't return anything on success

vladimir-shcherbakov replied:
the same comment.

vladimir-shcherbakov commented:
@talhers Ping on the comments above.

vladimir-shcherbakov commented:
@talhers  Ping on the comments above.

talhers commented:
Note on latest commit:

We discovered a bug where users with VA policy defined only at server level were not able to perform actions with our cmdlets (such as scan, baseline, export etc) because all VA cmdlets were checking that VA policy is defined at database level. This check is not done properly and is also redundant because it is performed at the backend.

Therefore I removed these checks from our cmdlets (remove SqlVulnerabilityAssessmentOperationCmdletBase class and remove method ApplyUserInputToModel from SqlVulnerabilityAssessmentAtpCmdletBase class).

These checks are performed in the backend so the cmdlets' behavior will remain the same.
I have updated the tests and added a case for checking this both on server and on managed instance.

yoavfr commented:
Looks good

vladimir-shcherbakov commented:
signing: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/psnetsign-esrp/250/

vladimir-shcherbakov commented on src/Sql/Sql/help/Update-AzSqlDatabaseVulnerabilityAssessmentSettings.md:
```diff
@@ -223,7 +223,7 @@ Parameter Sets: UpdateByInputObjectParameterSet
 Required: True
-Position: Named
+Position: 0
 Default value: None
```
I not sure why you touched this file but it's a breaking change.

talhers replied:
Fixed

vladimir-shcherbakov commented on src/Sql/Sql/help/Update-AzSqlInstanceDatabaseVulnerabilityAssessmentSettings.md:
```diff
@@ -223,7 +223,7 @@ Parameter Sets: UpdateByInputObjectParameterSet
 
 Required: True
-Position: Named
+Position: 0
```
I not sure why you touched this file but it's a breaking change.

talhers replied:
Fix
